### PR TITLE
HHH-17395 Refresh with PESSIMISTIC_WRITE ignored for lazy loaded entity HHH-1645 refresh with LockMode on an unitialized proxy does not work

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-1645
https://hibernate.atlassian.net/browse/HHH-17395

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java [hunk 1]
```diff
@@ -102,6 +104,19 @@ public final SharedSessionContractImplementor getSession() {
 		return session;
 	}
 
+	@Override
+	public final void setLockMode(LockMode lockMode) {
+		this.requestedLockMode = lockMode;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java:109]: Shouldn't this only be called when uninitialized?

```suggestion
		assert isUninitialized();
		this.requestedLockMode = lockMode;
```

File: hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java [hunk 2]
```diff
@@ -102,6 +104,19 @@ public final SharedSessionContractImplementor getSession() {
...
+	}
+
+	@Override
+	public final LockMode getLockMode() {
+		if ( requestedLockMode == null ) {
+			return LockMode.NONE;
+		}
+		return requestedLockMode;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java:117]: Shouldn't this only be called when uninitialized?

```suggestion
		assert isUninitialized();
		return requestedLockMode == null ? LockMode.NONE : requestedLockMode;
```

File: hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java [hunk 3]
```diff
@@ -173,6 +188,7 @@ else if ( !session.isConnected() ) {
 				else {
 					target = session.immediateLoad( entityName, id );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java:189]: Does the change you did in `SessionImpl#getCurrentLockMode` affect the SQL statement to include the lock mode in the select statement here? It should IMO, and if it does, then you shouldn't have to set the lock mode on the entity entry afterwards AFAIU.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java:189]: the select statement performed by the `immediateLoad` does not include the lock

**Comment by dreab8** [comment]
superseded by https://github.com/hibernate/hibernate-orm/pull/7611

**Pull request closed without merging**